cognitive/src/main/scala/com/microsoft/azure/synapse/ml/cognitive/anomaly/MultivariateAnomalyDetection.scala

```diff
@@ -284,61 +313,39 @@ trait MADBase extends HasOutputCol
     StorageInfo(account, uri.getUserInfo, key.get, uri.getPath.stripPrefix("/"))
   }
 
-  protected def blob: String = s"${getStorageInfo.blob}/$uid.zip"
+  protected def blob: String = s"${getStorageInfo.blob}/$uid.csv"
 
-  protected def blobPath: Path = new Path(new URI(getIntermediateSaveDir.stripSuffix("/") + s"/$uid.zip"))
+  protected def blobPath: Path = new Path(new URI(getIntermediateSaveDir.stripSuffix("/") + s"/$uid.csv"))
 
   protected def upload(df: DataFrame): String = {
-    val timestampColumn = df.schema
-      .find(p => p.name == getTimestampCol)
-      .get
-    val timestampColIdx = df.schema.indexOf(timestampColumn)
-    val rows = df.collect
-
+    val convertTimeFormatUdf = UDFUtils.oldUdf(
+      { value: String => convertTimeFormat("Timestamp column", value) },
+      StringType
+    )
+    val formatDf = df.withColumn(getTimestampCol, convertTimeFormatUdf(col(getTimestampCol)))
+      .sort(col(getTimestampCol).asc)
     val storageInfo = getStorageInfo
     val hconf = df.sparkSession.sparkContext.hadoopConfiguration
     val fs = FileSystem.get(blobPath.toUri, hconf)
 
-    // Upload to Blob
-    using(new ZipOutputStream(new BufferedOutputStream(fs.create(blobPath)))) { os =>
-      // loop over all features
-      for (feature <- df.schema.filter(p => p != timestampColumn).zipWithIndex) {
-        val featureIdx = df.schema.indexOf(feature._1)
-        // create zip entry. must be named series_{idx}
-        os.putNextEntry(new ZipEntry(s"series_${feature._2}.csv"))
-        // write CSV
-        storeFeatureInCsv(rows, timestampColIdx, featureIdx, os)
-        os.closeEntry()
-      }
-    }.get
-
-    // generate SAS
-    val offset = OffsetDateTime.now().plusDays(1)
-    val sas = SlimStorageClient.generateReadSAS(
-      storageInfo.account, storageInfo.container, blob, storageInfo.key, offset)
-    s"https://${storageInfo.account}.blob.core.windows.net/${storageInfo.container}/$blob?$sas"
-  }
-
-  protected def storeFeatureInCsv(rows: Array[Row],
-                                  timestampColIdx: Int,
-                                  featureIdx: Int,
-                                  out: OutputStream): Unit = {
-    // create CSV file per feature
-    val pw = new PrintWriter(out)
-    // CSV header
-    pw.println("timestamp,value")
-    for (row <- rows) {
-      // <timestamp>,<value>
-      // make sure it's ISO8601. e.g. 2021-01-01T00:00:00Z
-      val formattedTimestamp = convertTimeFormat("Timestamp column", row.getString(timestampColIdx))
-      pw.print(formattedTimestamp)
-      pw.write(',')
-      // TODO: do we have to worry about locale?
-      // pw.format(Locale.US, "%f", row.get(featureIdx))
-      pw.println(row.get(featureIdx))
+    val rows = formatDf.collect()
```

Comment:
We no longer need this I think you might be able to do this with a quick 

spark.coalesce(1).write.csv

Reply:
If we use spark to write to csv, it doesn't generate a single csv file, it still stores all data into xx.csv/part-00000 and in this case anomaly detector doesn't recognize it as a csv file input.

Reply:
Even if we pass xx.csv/part-0000 as the CSV instead of just xx.csv?

Reply:
You're right! This works!

Reply:
Thanks for trying 🙏